node/mock-answer.ts

```diff
@@ -22,18 +22,67 @@ export interface TaskLibAnswers {
     rmRF?: { [key: string]: { success: boolean } },
     stats?: { [key: string]: any }, // Can't use `fs.Stats` as most existing uses don't mock all required properties
     which?: { [key: string]: string },
+    variables?: {
+        nonSecrets?: { [key: string]: string },
```

Comment:
I considered adding a `fallbackToTask?: boolean` property, so consumers can specify whether they're specifying all defined variables or not. But mock-answer doesn't have access to task, and I didn't see a simple way to honor that.

Reply:
Makes sense, I don't think we need that.

Reply:
Naming suggestion: `plaintext`?

Reply:
Also, I think the types here should be

```ts
{ [key: string]: string | undefined }
```

We will want to test task behavior when some variable isn't set.

Reply:
`getVariable` will return `undefined` if the variable with name `name` isn't found. There's no fallback logic if variable answers were provided.

Does the real task allow for variables set to undefined? [`setVariable` takes a `string`](https://github.com/Microsoft/azure-pipelines-task-lib/blob/master/node/task.ts#L161) and [VariableInfo](https://github.com/Microsoft/azure-pipelines-task-lib/blob/master/node/task.ts#L204) has a `string` property. Wouldn't `getVariables` [blow up](https://github.com/Microsoft/azure-pipelines-task-lib/blob/master/node/task.ts#L149) if there were undefined values?

Reply:
Yeah, I believe if you `getVariable('SOME_UNDEFINED_ENVIRONMENT_VARIABLE')` the function will return undefined.

Reply:
Right. Why do you want to be able to _explicitly_ say:

```json
variables: {
   "plaintext": {
      ...
      "SOME_UNDEFINED_ENVIRONMENT_VARIABLE": undefined
   }
}
```

When the mocked `getVariable` for any variable not listed in the answers will return `undefined` anyway?

Reply:
Yeah, that's what I meant.  I'm under the impression that omitting the mock may or may not return undefined; it will fall through to the actual implementation and depend on whatever's in the process's environment

Reply:
I made a few updates. It should be clearer now that `getVariables` doesn't fall through.

```js
// variables not mocked.
answers: {}
```
and

```json
// variables mocked - getVariable returns undefined for any variable name.
answers: {
  "variables": {}
}
```